### PR TITLE
Benchmarking toolkit wrap up

### PR DESCRIPTION
**Pull request by abidwael**
This PR includes
- Enabling profiling with `LudwigProfiler`.
- README for the benchmarking toolkit.

### CODE REVIEW

File: ludwig/benchmarking/profiler_callbacks.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            use_torch_profiler=self.use_torch_profiler,
+            logging_interval=self.logging_interval,
+        )
+        self.train_profiler.__enter__()
+
+    def on_train_end(self, *args, **kwargs):
+        self.train_profiler.__exit__(None, None, None)
+        del self.train_profiler
+
+    def on_evaluation_start(self):
```
Comment (jppgks) [inline ludwig/benchmarking/profiler_callbacks.py:45]: probably need to consume args and kwargs here

File: ludwig/benchmarking/profiler_callbacks.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        self.evaluation_profiler = LudwigProfiler(
+            tag=EVALUATION,
+            output_dir=self.experiment_name,
+            use_torch_profiler=self.use_torch_profiler,
+            logging_interval=self.logging_interval,
+        )
+        self.evaluation_profiler.__enter__()
+
+    def on_evaluation_end(self):
```
Comment (jppgks) [inline ludwig/benchmarking/profiler_callbacks.py:54]: probably need to consume args and kwargs here

**Review by justinxzhao — COMMENTED** [review]
It looks like you're running into an obsolete broken test -- could you rebase?

File: ludwig/benchmarking/profiler_callbacks.py (hunk 1 above)
Comment (abidwael) [reply ludwig/benchmarking/profiler_callbacks.py:45]: I believe this callback doesn't take any args https://github.com/ludwig-ai/ludwig/blob/2ffd276976f882cab7baa7958e8cb43441b7bdc2/ludwig/callbacks.py#L331

File: ludwig/benchmarking/profiler_callbacks.py (hunk 2 above)
Comment (abidwael) [reply ludwig/benchmarking/profiler_callbacks.py:54]: same here, I think it doesn't take args https://github.com/ludwig-ai/ludwig/blob/2ffd276976f882cab7baa7958e8cb43441b7bdc2/ludwig/callbacks.py#L335

File: ludwig/benchmarking/utils.py [hunk 3]
```diff
@@ -146,6 +147,19 @@ def delete_model_checkpoints(output_directory: str):
         os.remove(os.path.join(output_directory, "model", "model_weights"))
 
 
+def delete_hyperopt_outputs(output_directory: str):
+    for path, currentDirectory, files in os.walk(output_directory):
+        for file in files:
+            filename = os.path.join(path, file)
+            if file not in ["hyperopt_statistics.json", "params.json", "stderr", "stdout", "result.json", "error.txt"]:
```
Comment (justinxzhao) [inline ludwig/benchmarking/utils.py:154]: nit: I would make the list a global constant at the top of the file, i.e.

```python
HYPEROPT_OUTDIR_RETAINED_FILES = ["hyperopt_statistics.json", "params.json", "stderr", "stdout", "result.json", "error.txt"]
```

File: requirements.txt [hunk 4]
```diff
@@ -29,6 +29,7 @@ torchinfo
 filelock
 psutil
 protobuf==3.20.1 # https://github.com/databrickslabs/dbx/issues/257
+s3fs
```
Comment (justinxzhao) [inline requirements.txt:32]: Maybe put this in a separate requirements file, i.e. `requirements_benchmarking.txt`

File: ludwig/benchmarking/summarize.py [hunk 5]
```diff
@@ -32,12 +36,73 @@ def summarize_metrics(
...
             resource_usage_diffs.append(resource_usage_diff)
         except Exception:
             logging.exception(f"Exception encountered while creating diff summary for {dataset_name}.")
-    return metric_diffs, resource_usage_diffs
+    shutil.rmtree(local_dir, ignore_errors=True)
+    export_and_print(dataset_list, metric_diffs, resource_usage_diffs)
+    return dataset_list, metric_diffs, resource_usage_diffs
+
+
+def export_and_print(
```
Comment (justinxzhao) [inline ludwig/benchmarking/summarize.py:52]: nit: Please add a docstring that describes what this function does -- it looks like there's a lot of interesting things happening with files getting saved and things getting logged.

File: ludwig/benchmarking/summarize.py [hunk 6]
```diff
@@ -32,12 +36,73 @@ def summarize_metrics(
...
+    dataset_list: List[str], metric_diffs: List[MetricsDiff], resource_usage_diffs: List[List[ResourceUsageDiff]]
+) -> None:
+    for dataset_name, experiment_metric_diff in zip(dataset_list, metric_diffs):
+        output_path = os.path.join("summarize_output", "performance_metrics", dataset_name)
+        os.makedirs(output_path, exist_ok=True)
+
+        logging.info(
```
Comment (justinxzhao) [inline ludwig/benchmarking/summarize.py:59]: `logger`?

File: ludwig/benchmarking/examples/process_config.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    ludwig_config["hyperopt"]["metric"] = dataset_name_to_metric[dataset_name]
+    ludwig_config["hyperopt"]["output_feature"] = ludwig_config["output_features"][0]["name"]
+
+    # use sparse encoder for categorical features to mimic logistic regression.
+    for i, feature in enumerate(ludwig_config["input_features"]):
+        if feature["type"] == "category":
+            ludwig_config["input_features"][i]["encoder"] = "sparse"
+    for i, feature in enumerate(ludwig_config["output_features"]):
+        if feature["type"] == "category":
+            ludwig_config["output_features"][i]["encoder"] = "sparse"
```
Comment (justinxzhao) [inline ludwig/benchmarking/examples/process_config.py:86]: Split this out into a separate function.

File: ludwig/benchmarking/examples/process_config.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
+# this function will take in a Ludwig config, strip away all its parameters except input and output features
```
Comment (justinxzhao) [inline ludwig/benchmarking/examples/process_config.py:1]: nit: Use `"""` docstring style for the top of python files.

File: ludwig/benchmarking/examples/process_config.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+# and add some other parameters to run logistic regression hyperopt.
+
+
+def process_config(ludwig_config: dict, experiment_dict: dict) -> dict:
+    """Modify a Ludwig config.
```
Comment (justinxzhao) [inline ludwig/benchmarking/examples/process_config.py:6]: Expand this docstring. How are the configs being processed?

**Review by justinxzhao — COMMENTED** [review]
Looks mostly good to me. Mainly just a few nits.

File: ludwig/benchmarking/examples/process_config.py (hunk 7 above)
Comment (abidwael) [reply ludwig/benchmarking/examples/process_config.py:86]: This file is just an example to demonstrate how to use it

**Comment by abidwael** [comment]
Using the `LudwigProfiler` as part of the callbacks is limited to `local` backend for now due to the following error.
```
Traceback (most recent call last):
  File "python/ray/_raylet.pyx", line 419, in ray._raylet.prepare_args_internal
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/serialization.py", line 433, in serialize
    return self._serialize_to_msgpack(value)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/serialization.py", line 411, in _serialize_to_msgpack
    pickle5_serialized_object = self._serialize_to_pickle5(
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/serialization.py", line 373, in _serialize_to_pickle5
    raise e
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/serialization.py", line 368, in _serialize_to_pickle5
    inband = pickle.dumps(
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 73, in dumps
    cp.dump(obj)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 620, in dump
    return Pickler.dump(self, obj)
TypeError: cannot pickle '_thread.lock' object

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ludwig/benchmarking/benchmark.py", line 122, in benchmark
    benchmark_one(experiment)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ludwig/benchmarking/benchmark.py", line 94, in benchmark_one
    _, _, _, output_directory = model.experiment(
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ludwig/api.py", line 1093, in experiment
    (train_stats, preprocessed_data, output_directory) = self.train(
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ludwig/api.py", line 557, in train
    train_stats = trainer.train(
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ludwig/backend/ray.py", line 386, in train
    results, self._validation_field, self._validation_metric = runner.run(
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/train/trainer.py", line 350, in run
    iterator = TrainingIterator(
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/train/trainer.py", line 683, in __init__
    self._start_training(
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/train/trainer.py", line 712, in _start_training
    self._run_with_error_handling(
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/train/trainer.py", line 722, in _run_with_error_handling
    return func()
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/train/trainer.py", line 713, in <lambda>
    lambda: self._backend_executor.start_training(
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/train/_internal/utils.py", line 185, in <lambda>
    return lambda *args, **kwargs: ray.get(actor_method.remote(*args, **kwargs))
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/actor.py", line 138, in remote
    return self._remote(args, kwargs)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/util/tracing/tracing_helper.py", line 425, in _start_span
    return method(self, args, kwargs, *_args, **_kwargs)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/actor.py", line 184, in _remote
    return invocation(args, kwargs)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/actor.py", line 171, in invocation
    return actor._actor_method_call(
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/actor.py", line 1164, in _actor_method_call
    object_refs = worker.core_worker.submit_actor_task(
  File "python/ray/_raylet.pyx", line 1730, in ray._raylet.CoreWorker.submit_actor_task
  File "python/ray/_raylet.pyx", line 1735, in ray._raylet.CoreWorker.submit_actor_task
  File "python/ray/_raylet.pyx", line 385, in ray._raylet.prepare_args_and_increment_put_refs
  File "python/ray/_raylet.pyx", line 376, in ray._raylet.prepare_args_and_increment_put_refs
  File "python/ray/_raylet.pyx", line 427, in ray._raylet.prepare_args_internal
TypeError: Could not serialize the argument <function construct_train_func.<locals>.<lambda> at 0x7f45c3d8cd30> for a task or actor ray.train._internal.backend_executor.BackendExecutor.start_training. Check https://docs.ray.io/en/master/ray-core/objects/serialization.html#troubleshooting for more information.
```

**Review by jppgks — APPROVED** [review]

**Comment by jppgks** [comment]
Debugging serializability of LudwigProfilerCallback:
```python
>>> from ray.util import inspect_serializability
>>> from ludwig.benchmarking.profiler_callbacks import LudwigProfilerCallback
>>> cb = LudwigProfilerCallback({'experiment_name': '/tmp', 'profiler': {'use_torch_profiler': False, 'logging_interval': 0}})

>>> inspect_serializability(cb, name="LudwigProfilerCallback")
================================================================================
Checking Serializability of <ludwig.benchmarking.profiler_callbacks.LudwigProfilerCallback object at 0xffff9d867b20>
================================================================================
(True, set())

>>> cb.on_preprocess_start()

>>> inspect_serializability(cb, name="LudwigProfilerCallback")
================================================================================
Checking Serializability of <ludwig.benchmarking.profiler_callbacks.LudwigProfilerCallback object at 0xffff9d6c4d90>
================================================================================
!!! FAIL serialization: cannot pickle '_thread.lock' object
    Serializing 'preload' <function Callback.preload at 0xffff7cdbf670>...
    Serializing '_abc_impl' <_abc_data object at 0xffff9d5b0300>...
    !!! FAIL serialization: cannot pickle '_abc_data' object
    WARNING: Did not find non-serializable object in <_abc_data object at 0xffff9d5b0300>. This may be an oversight.
================================================================================
Variable: 

        FailTuple(_abc_impl [obj=<_abc_data object at 0xffff9d5b0300>, parent=<ludwig.benchmarking.profiler_callbacks.LudwigProfilerCallback object at 0xffff9d6c4d90>])

was found to be non-serializable. There may be multiple other undetected variables that were non-serializable. 
Consider either removing the instantiation/imports of these variables or moving the instantiation into the scope of the function/class. 
If you have any suggestions on how to improve this error message, please reach out to the Ray developers on github.com/ray-project/ray/issues/
================================================================================
(False, {FailTuple(_abc_impl [obj=<_abc_data object at 0xffff9d5b0300>, parent=<ludwig.benchmarking.profiler_callbacks.LudwigProfilerCallback object at 0xffff9d6c4d90>])})
```

**Comment by abidwael** [comment]
@jppgks Thanks for investigating! That was the issue I encountered with a ray backend. I looked at other options of how we can still achieve this (using mutliprocessing instead of threading, etc.) but didn't get to a working solution. For now, let's just default to `local` backend but will be investigating this after this merge.

**Pull request merged**